### PR TITLE
[BUGFIX] - Broken Slave TWI Fixed

### PR DESCRIPTION
**Pull request by shahzadlone**
COMMIT LEGEND:
[Broken_C=173c3ebaa483d09acefb62df920bc6b064bf4420](https://github.com/mikaelpatel/Cosa/commit/173c3ebaa483d09acefb62df920bc6b064bf4420)
[Latest_C=90b815b6bf1a75af1e34b0f8f50ea6364e4d4acd](https://github.com/mikaelpatel/Cosa/commit/90b815b6bf1a75af1e34b0f8f50ea6364e4d4acd)
[ReBroke_C=a0d6795c04086423fb25fbd232c56a3e7ce9caa2](https://github.com/mikaelpatel/Cosa/commit/a0d6795c04086423fb25fbd232c56a3e7ce9caa2)

What i did:
1) Reverted back to the commit where things first broke at (<[Broken_C](https://github.com/mikaelpatel/Cosa/commit/173c3ebaa483d09acefb62df920bc6b064bf4420)>).
2) Fixed the bug (There was Nulling a variable and setting another to 0).
3) But even with my fix the latest commit (<[Latest_C](https://github.com/mikaelpatel/Cosa/commit/90b815b6bf1a75af1e34b0f8f50ea6364e4d4acd)>) was still broken, so i traced where it broke again after the fix. 
4) I found it rebroke after about 47 commits overall (and 2 commit changes after on the file TWI.cpp) at <[ReBroke_C](https://github.com/mikaelpatel/Cosa/commit/a0d6795c04086423fb25fbd232c56a3e7ce9caa2)>. I did manually merge all commits upto this commit to confirm everything works fine until this commit.
5) After that with a hint from @mikaelpatel, I was able to fix the second bug (needed to poweron TWI::slave:begin(). 
6) After that fixed i merge it up-to the Latest_C commit and everything seemed to work fine.

Will appreciate if this can be tested further to ensure i didn't break something else.
TODO: I commented the null and setting to 0, should i just delete them or leave them commented.

### CODE REVIEW

**Comment by shahzadlone** [comment]
Also might want to squash before merging to get rid of my nasty history of testing stuff in a weird way.

**Comment by shahzadlone** [comment]
@dhillondeep

**Comment by dhillondeep** [comment]
I will clean this up

**Review by mikaelpatel — CHANGES_REQUESTED** [review]
Removing these two lines is not a very good idea as they play an essential role in handling multi-threaded and asynchronous TWI calls. By removing them TWI drivers will block.

**Comment by dhillondeep** [comment]
@mikaelpatel so this means even though this fixes the issue, there is something else causing the problem? We should then keep these two lines of code and look for bug elsewhere in the code.

**Comment by mikaelpatel** [comment]
@dhillondeep yes, I would recommend removing these and rerunning the slave/master test sketches. If there is still an issue please let me know and I will review the test sketches.

**Comment by mikaelpatel** [comment]
@dhillondeep BW I see that the TWI example/test suite lacks sketches that use the asynchronous mode. It is a while (nearly 2 years) since I worked on that so I guess the recommendation is to avoid the asynchronous call-back variant and instead poll the completion if code needs to be run while waiting for the TWI device. Also please note that the TWI Slave will "hide" during the handling of a command block. It will remove will not answer (no address) until the response is completed and ready to be read by the Master. This is a trick that many TWI device use.

**Comment by dhillondeep** [comment]
This makes a lot of sense because we have been using the non-asynchronous variant for the whole time. Also thanks for the trick at the end it will help a lot with debugging. If @shahzadlone and I get time we will try creating an asynchronous example for TWI and will see how well that works! Until then we will resolve this bug and make a PR to the main repo.

**Comment by shahzadlone** [comment]
Made some changes to make both the async(i think) and sync work.

File: cores/cosa/Cosa/TWI.cpp [hunk 1]
```diff
@@ -292,6 +294,7 @@ TWI::Slave::begin()
   power_twi_enable();
 
   twi.m_dev = this;
+
```
Comment (Mogball) [inline cores/cosa/Cosa/TWI.cpp]: lol why

**Review by Mogball — CHANGES_REQUESTED** [review]

**Pull request merged**